### DISCUSSION TITLE
Why do boolean operations return `Nullable<Bool>` on Nullable values instead of `Bool`

### DISCUSSION THREAD

**Discussion by SteelAlloy** [diesel-rs/diesel#3462] [Q&A] [answered]
Hi there

Just a little question about nullable values and operators, I might have missed something.

Let's take a user table with a nullable name.
```rust
pub struct User {
  pub name: Option<String>,
}
...
users::table.filter(name.ilike("%test%")).get_result<User>()?;
```

Why does `name.ilike("%test%")` result in `Nullable<Bool>` instead of simply `Bool` ? I mean, the result will be `TRUE` or `FALSE` in the end, no ?

It's causing me some type issues with `BoxableExpression<SqlType = Bool>`, so I just want to know if I get something wrong or not.

Thanks a lot !

**Comment by weiznich** [comment, marked as answer]
> I mean, the result will be TRUE or FALSE in the end, no ?

That's exactly the assumption that's not true. You can test that by connecting to some postgres database via `psql` and execute the following commands:

```sql
\pset null '<NULL>` -- set the null escape sequence to see something in the following output
select NULL ILIKE 'foo%';
```

which results in the following output:

```
 ?column?
----------
 <NULL>
```

That means SQL operators return `TRUE`, `FALSE` or `NULL` depending on the operators values. So for cases where one of the values is `Nullable<_>` we must assume that the result is nullable as well. For where clauses database systems will normally coerce `NULL` values to false.

> It's causing me some type issues with BoxableExpression<SqlType = Bool>, so I just want to know if I get something wrong or not.

You can work around this "problem" by just making all values nullable via the [`NullableExpressionMethods::nullable`](https://docs.diesel.rs/master/diesel/expression_methods/trait.NullableExpressionMethods.html#method.nullable) method. `some_column.nullable().ilike("whatever")` will create a `BoxableExpression<SqlType = sql_types::Nullable<Bool>>` expression.

**Reply by SteelAlloy** [reply]
Thanks a lot, a clear and precise answer
